File: 1206.design-skiplist.java
#
# @lc app=leetcode id=1206 lang=java
#
# [1206] Design Skiplist
#

# @lc code=start
class Skiplist {

    private static final int MAX_LEVEL = 16;

    private Node head;
    private int maxLevel;

    private static class Node {
        int val;
        Node[] forward;
        Node(int val, int level) {
            this.val = val;
            forward = new Node[level];
        }
    }

    private int randomLevel() {
        int level = 1;
        while (Math.random() < 0.25 && level < MAX_LEVEL) {
            ++level;
        }
        return level;
    }

    public Skiplist() {
        head = new Node(-1, MAX_LEVEL);
        maxLevel = 1;
    }

    public boolean search(int target) {
        Node cur = head;
        for (int i = maxLevel - 1; i >= 0; --i) {
            while (cur.forward[i] != null && cur.forward[i].val < target) {
                cur = cur.forward[i];
            }
        }
        cur = cur.forward[0];
        return cur != null && cur.val == target;
    }

    public void add(int num) {
        Node[] update = new Node[MAX_LEVEL];
        Node cur = head;
        for (int i = maxLevel - 1; i >= 0; --i) {
            while (cur.forward[i] != null && cur.forward[i].val < num) {
                cur = cur.forward[i];
            }
            update[i] = cur;
        }
        int level = randomLevel();
        if (level > maxLevel) {
            for (int i = maxLevel; i < level; ++i) {
                update[i] = head;
            }
            maxLevel = level;
        }
        Node newNode = new Node(num, level);
        for (int i = 0; i < level; ++i) {
            newNode.forward[i] = update[i].forward[i];
            update[i].forward[i] = newNode;
        }
    }

    public boolean erase(int num) {
        Node[] update = new Node[MAX_LEVEL];
        Node cur = head;
        for (int i = maxLevel - 1; i >= 0; --i) {
            while (cur.forward[i] != null && cur.forward[i].val < num) {
                cur = cur.forward[i];
            }
            update[i] = cur;
        }
        Node cand = update[0].forward[0];
        if (cand == null || cand.val != num) {
            return false;
        }
        for (int i = 0; i < maxLevel; ++i) {
            if (update[i].forward[i] == cand) {
                update[i].forward[i] = cand.forward[i];
            }
        }
        return true;
    }
}

/**
 * Your Skiplist object will be instantiated and called as such:
 * Skiplist obj = new Skiplist();
 * boolean param_1 = obj.search(target);
 * obj.add(num);
 * boolean param_3 = obj.erase(num);
 */
# @lc code=end

File: 1206.design-skiplist.go
#
# @lc app=leetcode id=1206 lang=golang
#
# [1206] Design Skiplist
#

import (
	"math/rand"
)

const MAX_LEVEL = 16

type Node struct {
	val     int
	forward []*Node
}

type Skiplist struct {
	head *Node
}


func Constructor() Skiplist {
	head := &Node{
		val:     -1,
		forward: make([]*Node, MAX_LEVEL+1),
	}
	return Skiplist{head: head}
}


func randomLevel() int {
	level := 1
	for level < MAX_LEVEL && rand.Float64() < 0.5 {
		level++
	}
	return level
}

func (this *Skiplist) Search(target int) bool {
	curr := this.head
	for i := MAX_LEVEL; i >= 0; i-- {
		for curr.forward[i] != nil && curr.forward[i].val < target {
			curr = curr.forward[i]
		}
	}
	next := curr.forward[0]
	return next != nil && next.val == target
}

func (this *Skiplist) Add(num int) {
	update := make([]*Node, MAX_LEVEL+1)
	curr := this.head
	for i := MAX_LEVEL; i >= 0; i-- {
		for curr.forward[i] != nil && curr.forward[i].val < num {
			curr = curr.forward[i]
		}
		update[i] = curr
	}
	level := randomLevel()
	newnode := &Node{
		val:     num,
		forward: make([]*Node, level+1),
	}
	for i := 0; i <= level; i++ {
		newnode.forward[i] = update[i].forward[i]
		update[i].forward[i] = newnode
	}
}

func (this *Skiplist) Erase(num int) bool {
	update := make([]*Node, MAX_LEVEL+1)
	curr := this.head
	for i := MAX_LEVEL; i >= 0; i-- {
		for curr.forward[i] != nil && curr.forward[i].val < num {
			curr = curr.forward[i]
		}
		update[i] = curr
	}
	if update[0].forward[0] == nil || update[0].forward[0].val != num {
		return false
	}
	for i := 0; i <= MAX_LEVEL; i++ {
		if update[i].forward[i] != nil && update[i].forward[i].val == num {
			update[i].forward[i] = update[i].forward[i].forward[i]
		}
	}
	return true
}


/**
 * Your Skiplist object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Search(target);
 * obj.Add(num);
 * param_3 := obj.Erase(num);
 */
# @lc code=end